operations/host_download.go

```diff
@@ -0,0 +1,164 @@
+)
+
+const (
+	modeLength   = 3
+	maxModeValue = 0777
+)
+
+// hostDownload creates a command to download a file from a URL and extract it
+// if desired.
+func hostDownload() cli.Command {
+	const (
+		destDirFlagName  = "dir"
+		extractFlagName  = "extract"
+		fileNameFlagName = "file"
+		modeFlagName     = "mode"
+		urlFlagName      = "url"
+	)
+	return cli.Command{
+		Name:  "download",
+		Usage: "download resources from a URL on a host and optionally extract",
+		Flags: []cli.Flag{
+			cli.StringFlag{
+				Name:  joinFlagNames(urlFlagName, "u"),
+				Usage: "URL to download",
+			},
+			cli.StringFlag{
+				Name:  joinFlagNames(destDirFlagName, "d"),
+				Usage: "destination directory for the resource",
+			},
+			cli.StringSliceFlag{
+				Name:  joinFlagNames(fileNameFlagName, "f"),
+				Usage: "the name of the output file in the destination directory (can be multiple if extracting)",
+			},
+			cli.StringFlag{
+				Name:  joinFlagNames(extractFlagName, "e"),
+				Usage: "extract the resource and put in the destination directory",
+			},
+			cli.StringFlag{
+				Name:  joinFlagNames(modeFlagName, "m"),
+				Usage: "permission bits for the files (if unset, no change)",
+			},
+		},
+		Before: mergeBeforeFuncs(
+			requireStringFlag(urlFlagName),
+			requireStringFlag(destDirFlagName),
+			requireStringSliceFlag(fileNameFlagName),
+		),
+		Action: func(c *cli.Context) error {
+			url := c.String(urlFlagName)
+			destDir := c.String(destDirFlagName)
+			fileNames := c.StringSlice(fileNameFlagName)
+			extract := c.Bool(extractFlagName)
+			mode, err := validateAndParseMode(c.String(modeFlagName))
+			if err != nil {
+				return errors.Wrap(err, "error parsing mode")
+			}
+
+			if !extract && len(fileNames) != 1 {
+				return errors.New("must specify exactly one file name if not extracting")
+			}
+
+			filePaths := make([]string, 0, len(fileNames))
+			for _, fileName := range fileNames {
+				filePaths = append(filePaths, filepath.Join(destDir, fileName))
+			}
+
+			if err := os.MkdirAll(destDir, 0755); err != nil {
+				return errors.Wrap(err, "error making destination directory")
+			}
+
+			if extract {
+				// If extracting, must extract to a directory.
+				if err := downloadAndExtract(url, destDir); err != nil {
+					return errors.Wrap(err, "error occurred during download and extract")
+				}
+			} else {
+				// If not extracting, just download directly into the file.
+				if err := download(url, filePaths[0]); err != nil {
+					return errors.Wrap(err, "error occurred during file download")
+				}
+			}
+
+			if mode != 0 {
+				if err := setFileModes(filePaths, mode); err != nil {
+					return errors.Wrap(err, "error occurred while changing file modes")
+				}
+			}
+
+			return nil
+		},
+	}
+}
+
+// downloadAndExtract downloads the archive file from url as a temporary
+// file, extracts the archive file to the directory destDir, and removes the
+// temporary archive.
+func downloadAndExtract(url, destDir string) error {
```

Comment:
doesn't jasper have support for this kind of operation? does it make more sense to delegate to that code?